File: python/CaptureService.py
# https://github.com/Chr1ll3D/real-time-whiteboard-app-main/blob/2df884466eac2a3e5848b14fc2db9682afc5f45f/app/src/main/java/com/whiteboardapp/core/CaptureService.java
from segmentator import segmentator
from binarization import binarization
from changeDetector import changeDetector

import numpy as np
import time
import cv2

class CaptureService:

    TAG = "CaptureService"

    def __init__(self, defaultWidth:int, defaultHeight:int):
        self.currentModel:np.ndarray = np.ones((defaultWidth, defaultHeight), dtype=np.uint8)*255
        self.changeDetector = changeDetector((defaultWidth, defaultHeight))
        self.binarization = binarization()
    

    # Runs image through the image processing pipeline
    def capture(self, imgBgr:np.ndarray) -> np.ndarray:

        # Segmentation
        print(f"{self.TAG} capture: Segmentation started.")
        matPerspectiveRgb:np.ndarray = cv2.cvtColor(imgBgr, cv2.COLOR_BGR2RGB)
        deeplab = segmentator()
        #Bitmap bitmapRgb = MatConverter.matToBitmap(matPerspectiveRgb)
        imgSegMap:np.ndarray = deeplab.segmentate(matPerspectiveRgb)
        print(f"{self.TAG} capture: Segmentation done.")
        cv2.imshow("imgSegMap", imgSegMap)

        # Binarize a gray scale version of the image.
        imgWarpGray:np.ndarray = cv2.cvtColor(imgBgr, cv2.COLOR_BGR2GRAY)
        imgBinarized:np.ndarray = self.binarization.binarize(imgWarpGray)
        cv2.imshow("imgBinarized", imgBinarized)

        # Remove segments before change detection.
        currentModelCopy, imgBinarizedcopy = self.removeSegmentArea(imgBinarized, imgSegMap)
        cv2.imshow("currentModelCopy", currentModelCopy)
        cv2.imshow("imgBinarizedcopy", imgBinarizedcopy)
        # Change detection
        imgPersistentChanges:np.ndarray = self.changeDetector.detectChanges(imgBinarizedcopy, currentModelCopy)
        cv2.imshow("imgPersistentChanges", imgPersistentChanges)
        # Update current model with persistent changes.
        self.updateModel(imgBinarizedcopy, imgPersistentChanges)
        return self.currentModel

    

    # Removes segment area from image
    def removeSegmentArea(self, binarizedImg:np.ndarray, imgSegMap:np.ndarray):
        # currentModelCopy:np.ndarray = self.currentModel.copy()

        startTime = (time.time() / 1000)

        # bufferBinarized = binarizedImg.copy()
        # bufferSegmap = imgSegMap.copy()
        # bufferModel = currentModelCopy.copy()

        # for i in range(len(bufferBinarized)):
        #     if (bufferSegmap[i] == -1):
        #         bufferBinarized[i] = -1
        #         bufferModel[i] = -1

        # binarizedImg.put(0, 0, bufferBinarized)
        # currentModelCopy.put(0, 0, bufferModel)

        imgBinarizedcopy = cv2.bitwise_and(binarizedImg, cv2.bitwise_not(imgSegMap))
        currentModelCopy = cv2.bitwise_and(self.currentModel, imgSegMap)

        endTime = (time.time() / 1000)
        print(f"Remove segment loop took: {(endTime - startTime)} milliseconds")
        return currentModelCopy, imgBinarizedcopy
    

    def updateModel(self, imgBinarized:np.ndarray, imgPersistentChanges:np.ndarray) -> None:

        startTime = (time.time() / 1000)

        # bufferModel = self.currentModel.copy()
        # bufferChanges = imgPersistentChanges.copy()
        # bufferBinarized = imgBinarized.copy()

        # for i in range(len(bufferModel)):
        #     if (bufferChanges[i] == -1):
        #         bufferModel[i] = bufferBinarized[i]

        # self.currentModel.put(0, 0, bufferModel)

        self.currentModel = cv2.bitwise_and(self.currentModel, imgPersistentChanges)
        imgBinarized = cv2.bitwise_and(imgBinarized, cv2.bitwise_not(imgPersistentChanges))
        self.currentModel = cv2.bitwise_or(self.currentModel, imgBinarized)

        endTime = (time.time() / 1000)
        print(f"Update model loop took: {(endTime - startTime)} milliseconds")

File: app/src/main/cpp/CaptureService.cpp
#include "binarization.hpp"
#include "changeDetector.hpp"
#include "segmentator.hpp"
#include "perspectiveTransformer.hpp"
#include <chrono>
#include <android/log.h>

using std::chrono::duration;
using clk = std::chrono::high_resolution_clock;

class CaptureService {
private:
    const char* TAG = "CaptureService";
    cv::Mat currentModel;
    cv::Size size;
    changeDetector ChangeDetector = changeDetector(cv::Size_<int>());
    binarization Binarization{};
    segmentator Segmentator{};
    int rounds = 0;
    double totalSegmentatorTime = 0;
    double totalBinarizationTime = 0;
    double totalRemoveTime = 0;
    double totalChangeTime = 0;
    double totalUpdateTime = 0;
public:
    explicit CaptureService(const cv::Size_<int> size): size(size) {
        currentModel = cv::Mat::ones(size, CV_8UC1) * 255;
        ChangeDetector = changeDetector(size);
    }

    explicit CaptureService(const cv::Size_<int>& size, const std::string& path) {
        currentModel = cv::Mat::ones(size, CV_8UC1) * 255;
        ChangeDetector = changeDetector(size);
        Segmentator = segmentator{path};
    }

    cv::Mat capture(const cv::Mat& imgBgr) {
        // Segmentation
        __android_log_print(ANDROID_LOG_DEBUG, "capture", "Beginning segmentation");
        rounds++;
        auto t0 = clk::now();
        cv::Mat matPerspectiveRgb;
        cv::cvtColor(imgBgr, matPerspectiveRgb, cv::COLOR_RGBA2RGB);
        cv::resize(matPerspectiveRgb, matPerspectiveRgb, cv::Size(480, 270));
        cv::Mat imgSegMap = Segmentator.segmentate(matPerspectiveRgb);
        cv::resize(imgSegMap, imgSegMap, size);
        auto t1 = clk::now();
        totalSegmentatorTime += duration<double, std::milli>(t1-t0).count();
        __android_log_print(ANDROID_LOG_DEBUG, "segmentate", "%s ms", std::to_string(totalSegmentatorTime/rounds).c_str());

        // Binarize a gray scale version of the image.
        t0 = clk::now();
        cv::Mat imgWarpGray;
        cv::cvtColor(imgBgr, imgWarpGray, cv::COLOR_RGBA2GRAY);
        cv::Mat imgBinarized = Binarization.binarize(imgWarpGray);
        t1 = clk::now();
        totalBinarizationTime += duration<double, std::milli>(t1-t0).count();
        __android_log_print(ANDROID_LOG_DEBUG, "binarize", "%s ms", std::to_string(totalBinarizationTime/rounds).c_str());

        // Remove segments before change detection.
        t0 = clk::now();
        cv::Mat currentModelCopy = removeSegmentArea(imgBinarized, imgSegMap);
        t1 = clk::now();
        totalRemoveTime += duration<double, std::milli>(t1-t0).count();
        __android_log_print(ANDROID_LOG_DEBUG, "removeSegmentArea", "%s ms", std::to_string(totalRemoveTime/rounds).c_str());

        // Change detection
        t0 = clk::now();
        cv::Mat imgPersistentChanges = ChangeDetector.detectChanges(imgBinarized, currentModelCopy);
        t1 = clk::now();
        totalChangeTime += duration<double, std::milli>(t1-t0).count();
        __android_log_print(ANDROID_LOG_DEBUG, "detectChanges", "%s ms", std::to_string(totalChangeTime/rounds).c_str());

        // Update current model with persistent changes.
        t0 = clk::now();
        updateModel(imgBinarized, imgPersistentChanges);
        t1 = clk::now();
        totalUpdateTime += duration<double, std::milli>(t1-t0).count();
        __android_log_print(ANDROID_LOG_DEBUG, "updateModel", "%s ms", std::to_string(totalUpdateTime/rounds).c_str());
        return currentModel;
    }

    // Removes segment area from image
    cv::Mat removeSegmentArea(cv::Mat& imgBinarized, cv::Mat& imgSegMap) { //TODO something seems wrong
        cv::Mat currentModelCopy = currentModel.clone();
        imgBinarized &= ~imgSegMap;
        currentModelCopy &= imgSegMap;
        return currentModelCopy;
    }

    void updateModel(cv::Mat imgBinarized, const cv::Mat& imgPersistentChanges) {
        currentModel = currentModel & imgPersistentChanges;
        imgBinarized = imgBinarized & ~imgPersistentChanges;
        currentModel = currentModel | imgBinarized;
    }
};